src/ir/ir_traversing_visitor.h

```diff
@@ -24,47 +26,89 @@ namespace raksha::ir {
 // A visitor that also traverses the children of a node and allows performing
 // different actions before (PreVisit) and after (PostVisit) the children are
 // visited. Override any of the `PreVisit` and `PostVisit` methods as needed.
-template <typename Derived>
-class IRTraversingVisitor : public IRVisitor<Derived> {
+template <typename Derived, typename Result = Unit>
+class IRTraversingVisitor : public IRVisitor<Derived, Result> {
  public:
   virtual ~IRTraversingVisitor() {}
 
+  // Gives a default value for all 'PreVisit's to start with.
+  // Should be over-ridden if the Result is not default constructable.
+  virtual Result GetDefaultValue() { return Result(); }
+
   // Invoked before all the children of `module` is visited.
-  virtual void PreVisit(const Module& module) {}
+  virtual Result PreVisit(const Module& module) {
+    return std::move(GetDefaultValue());
```

Comment:
`std::move` is not needed as `GetDefaultValue()` does not have a name. I just remembered  https://abseil.io/tips/77. It is a good read and gives a simple mental model for `std::move`.

Reply:
Gotcha, thanks